pkg/executables/cmk.go

```diff
@@ -0,0 +1,265 @@
+func (c *Cmk) ValidateAccountPresent(ctx context.Context, account string) error {
+	result, err := c.exec(ctx, "list", "accounts", fmt.Sprintf("name=\"%s\"", account))
+	if err != nil {
+		return fmt.Errorf("error getting accounts info: %v", err)
+	}
+	if result.Len() == 0 {
+		return fmt.Errorf("account %s not found", account)
+	}
+
+	response := struct {
+		CmkAccounts []CmkAccount `json:"account"`
+	}{}
+	if err = json.Unmarshal(result.Bytes(), &response); err != nil {
+		return fmt.Errorf("failed to parse response into json: %v", err)
+	}
+	accounts := response.CmkAccounts
+	if len(accounts) > 1 {
+		return fmt.Errorf("duplicate account %s found", account)
+	} else if len(accounts) == 0 {
+		return fmt.Errorf("account %s not found", account)
+	}
+	return nil
+}
+
+func NewCmk(executable Executable, writer filewriter.FileWriter, config CmkExecConfig) *Cmk {
+	return &Cmk{
+		writer:     writer,
+		executable: executable,
+		retrier:    retrier.NewWithMaxRetries(maxRetriesCmk, backOffPeriodCmk),
+		config:     config,
+	}
+}
+
+// ValidateCloudStackConnection Calls `cmk sync` to ensure that the endpoint and credentials + domain are valid
+func (c *Cmk) ValidateCloudStackConnection(ctx context.Context) error {
+	buffer, err := c.exec(ctx, "sync")
+	if err != nil {
+		return fmt.Errorf("error validating cloudstack connection for cmk config %s: %v", buffer.String(), err)
+	}
+	logger.MarkPass("Connected to CloudStack server")
+	return nil
+}
+
+func (c *Cmk) exec(ctx context.Context, args ...string) (stdout bytes.Buffer, err error) {
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed get environment map: %v", err)
+	}
+	configFile, err := c.buildCmkConfigFile()
+	if err != nil {
+		return bytes.Buffer{}, fmt.Errorf("failed cmk validations: %v", err)
+	}
+	argsWithConfigFile := append([]string{"-c", configFile}, args...)
+
+	return c.executable.Execute(ctx, argsWithConfigFile...)
+}
+
+// TODO: Add support for passing in domain from Deployment Config Spec
+type CmkExecConfig struct {
+	CloudStackApiKey        string // Api Key for CloudMonkey to access CloudStack Cluster
+	CloudStackSecretKey     string // Secret Key for CloudMonkey to access CloudStack Cluster
+	CloudStackManagementUrl string // Management Endpoint Url for CloudMonkey to access CloudStack Cluster
+	CloudMonkeyVerifyCert   bool   // boolean indicating if CloudMonkey should verify the cert presented by the CloudStack Management Server
+}
+
+func (c *Cmk) buildCmkConfigFile() (configFile string, err error) {
+	t := templater.New(c.writer)
+	writtenFileName, err := t.WriteToFile(cmkConfigTemplate, c.config, cmkConfigFileName)
+	if err != nil {
+		return "", fmt.Errorf("error creating file for cmk config: %v", err)
+	}
+	configFile, err = filepath.Abs(writtenFileName)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate absolute filepath for generated config file at %s", writtenFileName)
+	}
+
+	return configFile, nil
+}
+
+type CmkTemplate struct {
```

Comment:
nit: did you get to try and make these structs private but keep the fields public?

Reply:
I did try that, and deserialization failed :(